Feat/#47 modal

soralee2821 opened the pull request:
## Issues
- Issue number #47 

## Tasks Done 
- [x] **createPortal** API를 사용하여 `Modal` 컴포넌트 구현
- [x] `Modal` 컴포넌트 안에 들어갈 `Modal Button`, `Prompt Text` 컴포넌트 구현

## Description
- 형식이 정해진 라이브러리를 사용하는 대신 React의 **createPortal** API를 사용하여 범용적인 Modal 컴포넌트를 구현했습니다.
  - `Modal` 컴포넌트의 **children** props를 받아서 자식 컴포넌트들을 합성하여 사용할 수 있도록 했습니다.
  - `Modal` 컴포넌트의 배경 컴포넌트의 **opacity**로 0.5를 적용했습니다.
  - `Modal` 컴포넌트의 자식 컴포넌트들에게 border-radius가 적용되도록 **overflow**로 hidden을 적용했습니다.
  - "네" 버튼 클릭 시 이동하는 경로를 `/delete-account`로 수정했습니다.
- `ModalButton` 컴포넌트를 구현했습니다.
  - **text**, **handler** props를 받아서 여러 버튼 컴포넌트로 재사용할 수 있도록 구현했습니다.
  - **imageUrl** props를 받을 경우, `ModalButton` 컴포넌트의 자식 컴포넌트로 **img** 요소를 렌더링하도록 했습니다.
  - **isPrimary** props의 값으로 true를 받을 경우, `ModalButton` 컴포넌트의 배경색이 메인색이 되도록 했습니다.
  - `ModalButton` 컴포넌트에 **min-height**로 40px(2.5rem) , **min-width**로 60px(3.75rem), **padding**로 8px(p-2)을 설정했습니다.
- `ModalText` 컴포넌트를 구현했습니다.
  - `ModalText` 컴포넌트의 **children** props를 받아서 자식 컴포넌트들을 합성하여 사용할 수 있도록 했습니다.
  - `ModalText` 컴포넌트에 **max-width**로 320px(20rem)을 설정했습니다.
  - `MainText` 컴포넌트에 **font-size**로 16px(text-base), **font-weight**로 800(font-extrabold)를 설정했습니다.
  - `SubText` 컴포넌트에  **font-size**로 14px(text-sm), **line-height**로 24px(leading-6)를 설정했습니다.

## Visuals

<img width="213" alt="스크린샷 2023-07-21 131535" src="https://github.com/Pullanner/pullanner-web/assets/70058081/8d003f97-b6c1-45ca-a9c6-04014d907e31">

soralee2821 committed:
[#47] Feat: add Modal component

- in order to implement modal, use React.createPortal
- add portal-root in index.html

soralee2821 committed:
[#47] Feat: add ModalButton component

- imageUrl, isFocused is optional parameter
- when isFocused is true, button is focused by using ref
- when imageUrl exist, button include img tag

soralee2821 committed:
[#47] Feat: add PromptText component

- add whitespace-normal css proprety to make multi-line text

soralee2821 committed:
[#47] Feat: add logic to show modal when withdrawal button clicked

soralee2821 committed:
[#47] Refactor: change to span tag and delete whitespace-normal CSS property

soralee2821 committed:
[#47] Refactor: add withdrawal path in ROUTE_PATH and apply to EditMyPage

jaypedia commented on src/components/Modal/ModalButton.tsx:
```diff
@@ -0,0 +1,30 @@
+    if (isFocused && buttonRef.current) {
+      buttonRef.current.focus();
+    }
+  }, [isFocused]);
```
useEffect 대신에 useLayoutEffect를 쓰신 이유도 PR에 포함해주실 수 있나요?

그리고 formatting에 관련한 이야긴데 return문과 한 줄 띄워서 작성하면 가독성이 좀 더 좋아질 것 같습니다. 저는 보통 컴포넌트 내에 함수 같은 것을 사용한다면 return문 사이에 공백 빈 줄을 넣거든요. 상의 후 컨벤션으로 지정해도 좋을 것 같아요~

soralee2821 replied:
모달이 띄워졌을 때 '아니오' 버튼에 포커스가 되어 있어야 하기 때문에, 마운트가 된 후 실행되는 useEffect 대신에 마운트 전에 실행되는 useLayoutEffect를 사용했습니다.
return 사이에 공백 넣는 거 좋습니다.

jaypedia replied:
padding-line-between-statements를 적용하면 되겠네요! 

https://eslint.org/docs/latest/rules/padding-line-between-statements

jaypedia commented on src/constants/routePath.ts:
```diff
@@ -17,4 +17,5 @@ export const ROUTE_PATH = {
   },
   summary: '/summary',
   community: '/community',
+  withdrawal: '/withdrawal',
```
withdrawal은 출금 느낌이 나서, 회원 탈퇴 페이지와는 조금 어울리지 않는다는 생각이 들었습니다.
'계정 삭제'를 의미하는 delete account로 변경하면 좋을 것 같네요!

soralee2821 replied:
네, delete-account로 변경하겠습니다.

jaypedia reviewed (changes requested):
Modal 컴포넌트를 합성 컴포넌트로 하면 좋겠다는 제안을 잘 받아주시고 구현을 잘 해주셨네요! 고생 많으셨어요👍 논의할 부분만 좀 더 논의 후 바로 머지하면 될 것 같아요.

추가적으로 Modal이 띄워졌을 때 Background가 Dim 처리 되는 것을 추가해야 할 것 같습니다.

soralee2821 committed:
[#47] Design: add background opacity in Modal

soralee2821 committed:
[#47] Refactor: modify withdrawal to deleteAccount

soralee2821 committed:
[#47] Refactor: apply padding-line-between-statements in ModalButton

soralee2821 committed:
[#47] Refactor: delete fixed values of width, height in Modal

jaypedia commented on src/components/Modal/index.tsx:
```diff
@@ -0,0 +1,20 @@
+export const Modal = ({ children }: ModalProps) => {
+  return createPortal(
+    <div className="fixed left-0 top-0 z-50 flex h-screen w-screen items-center justify-center">
+      <div className="flex h-[100%] w-[100%] items-center justify-center bg-black/50 sm:h-[50rem] sm:max-h-[90vh] sm:w-96">
```
bg-black/50 이런 식으로도 opacity를 조절할 수 있다는 걸 알았네요! 

https://tailwindcss.com/docs/background-color#changing-the-opacity

여기서도 나오는군요 ㅎㅎ 덕분에 알아갑니다
이부분이 모달의 background가 dim 처리 된 부분인가요?
dim 처리 하셨다고 한 것 같은데, 커밋에서는 보이지 않네요! 커밋 메시지에도 이러한 변경 사항은 남겨주시면 좋겠습니다.

더불어 dim 처리 된 것이 반영될 수 있도록 Visuals에 추가해주시면 좋을 것 같습니다!

soralee2821 replied:
네, bg-black/50이 opacity 0.5로 수정한 부분 맞습니다. Visuals 반영해서 변경할께요.

jaypedia commented on src/components/Modal/index.tsx:
```diff
@@ -10,7 +10,7 @@ export const Modal = ({ children }: ModalProps) => {
     <div className="fixed left-0 top-0 z-50 flex h-screen w-screen items-center justify-center">
       <div className="flex h-[100%] w-[100%] items-center justify-center bg-black/50 sm:h-[50rem] sm:max-h-[90vh] sm:w-96">
-        <div className="z-50 flex h-40 w-[21.375rem] flex-col items-center justify-center rounded-md bg-[#202020] p-5">
+        <div className="z-50 flex flex-col items-center justify-center rounded-md bg-[#202020] p-5">
```
고정값으로 주셨던 height, width를 제거하셨네요 굿!

jaypedia commented on src/pages/EditMyPage/index.tsx:
```diff
@@ -45,7 +61,31 @@ export const EditMyPage = () => {
+      {showModal && (
+        <Modal>
+          <PromptText>
+            <p className="mb-2.5 text-xs">
```
모달 버튼의 폰트 크기는 text-sm인데, prompt text의 크기는 text-xs네요! 버튼 안의 텍스트가 더 큰 것이 시각적으로 약간 어색해 보이기도 합니다. 

![image](https://github.com/Pullanner/pullanner-web/assets/85419343/bf84ddea-efb5-4831-9480-257480502f3b)

아니면 위와 같은 예시로 메인 텍스트는 사이즈를 좀 더 크게 하고, 서브 텍스트는 크기를 작게 하면 UI가 좀 더 자연스러우면서도 유저에게 잘 알릴 수 있을 것 같아요.
'정말 탈퇴하시겠습니까?'를 메인 텍스트로, 그 외의 문구를 서브 텍스트로 하면 어떨까요?

soralee2821 replied:
PromptText의 글자 크기가 **text-xs**인 이유는 "회원 탈퇴 시 Pullanner의 모든 데이터가 삭제됩니다." 문구에서 마지막 글자인 "다"만 다음 줄로 넘어가는 것이 미관 상 안좋아보여서 글자 크기를 줄인 것이었습니다.
그런데 PromptText의 글자 크기가 ModalButton의 글자 크기보다 작으면 안 되겠네요. 반영해서 수정하겠습니다.

soralee2821 replied:
MainText와 SubText를 나눠서 구현하는 방법 좋습니다!

soralee2821 committed:
[#47] Refactor: apply PR reviews in Modal

- PromptText를 ModalText로 파일명 변경
- ModalText를 MainText와 SubText로 분리
- MainText일 때 text-base, font-extrabold CSS 추가
- SubText일 때 mb-0.5, text-sm CSS 추가
- "정말 탈퇴하시겠습니까?" 문구에 text-red-400 CSS 추가
- ModalButton에서 focus 관련 로직 삭제
- ModalButton에서 isPrimary가 true일 때 메인 색으로 변경되는 로직 추가
- ModalButton에서 고정 width, height를 min-width, min-height로 변경
- Modal에서 boader-radius를 자식 컴포넌트에 모두 적용하기 위해 overflow: hidden CSS 추가

Co-authored-by: Jayun Lee <jayunlee.dev@gmail.com>

soralee2821 committed:
[#47] Refactor: rename PromptText to ModalText component in ModalText

soralee2821 committed:
[#47] Refactor: add max-width to ModalText

soralee2821 committed:
[#47] Refactor: add line-height to SubText

soralee2821 committed:
[#47] Refactor: remove margin-bottom in SubText

soralee2821 committed:
[#47] Refactor: add margin-bottom to button container div

jaypedia commented on src/pages/EditMyPage/index.tsx:
```diff
@@ -70,7 +70,7 @@ export const EditMyPage = () => {
             </SubText>
           </ModalText>
-          <div className="flex w-full justify-around p-3">
+          <div className="mb-2 flex w-full justify-around p-3">
```
확실피 마진 추가하니까 UI가 낫네요! 👍
